server/console_channel.go

```diff
@@ -0,0 +1,113 @@
+package server
+
+import (
+	"context"
+	sql "database/sql"
+	"fmt"
+	"github.com/gofrs/uuid"
+	"github.com/heroiclabs/nakama-common/api"
+	"github.com/heroiclabs/nakama-common/runtime"
+	"github.com/heroiclabs/nakama/v3/console"
+	"github.com/jackc/pgtype"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+	"net/url"
+	"sort"
+	"time"
+)
+
+func (s *ConsoleServer) ListChannelMessages(ctx context.Context, in *console.ListChannelMessagesRequest) (*api.ChannelMessageList, error) {
+	const limit = 50
+
+	stream, err := buildStream(in)
+	if err != nil {
+		return nil, err
+	}
+
+	channelId, err := StreamToChannelId(*stream)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	cursor, err := url.QueryUnescape(in.Cursor)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, "Cursor is invalid or expired.")
+	}
+
+	messageList, err := ChannelMessagesList(ctx, s.logger, s.db, uuid.Nil, *stream, channelId, limit, false, cursor)
+	if err == runtime.ErrChannelCursorInvalid {
+		return nil, status.Error(codes.InvalidArgument, "Cursor is invalid or expired.")
+	} else if err != nil {
+		return nil, status.Error(codes.Internal, "Error listing messages from channel.")
+	}
+
+	return messageList, nil
+}
+
+func (s *ConsoleServer) DeleteChannelMessage(ctx context.Context, in *console.DeleteChannelMessageRequest) (*emptypb.Empty, error) {
+	messageID, err := uuid.FromString(in.Id)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, "Requires a valid message ID.")
+	}
+
+	query := "DELETE FROM message WHERE id = $1::UUID"
+	if _, err := s.db.ExecContext(ctx, query, messageID); err != nil {
+		s.logger.Debug("Could not delete message.", zap.Error(err))
+		return nil, status.Error(codes.Internal, "An error occurred while trying to delete the message.")
+	}
+
+	s.logger.Info("Message deleted.", zap.String("message_id", messageID.String()))
+	return &emptypb.Empty{}, nil
+}
+
+func (s *ConsoleServer) DeleteOldChannelMessages(ctx context.Context, in *console.DeleteOldChannelMessageRequest) (*console.MessagesDeleted, error) {
+	query := "DELETE FROM message WHERE create_time < $1::TIMESTAMPTZ"
+	deleteBefore := time.Unix(in.DeleteBefore.Seconds, int64(in.DeleteBefore.Nanos)).UTC()
+
+	var res sql.Result
+	var err error
+	if res, err = s.db.ExecContext(ctx, query, &pgtype.Timestamptz{Time: deleteBefore, Status: pgtype.Present}); err != nil {
+		s.logger.Debug("Could not delete old messages.", zap.Error(err))
+		return nil, status.Error(codes.Internal, "An error occurred while trying to delete old message.")
+	}
+	affected, err := res.RowsAffected()
+	if err != nil {
```

Comment:
I think in this case just log the error and return 0 deleted messages in the response. The old messages have actually been deleted, so returning an error doesn't quite feel right.